### PR TITLE
Builtin processor for unpivoting Tabular data

### LINKED ISSUE #108: Unpivot  tabular data
As a dpp user, I want to unpivot or transpose my tabular data so that there's only one record per row

As a dpp user, I have data that has dates as  column names and I want to "unpivot" and have nice time series data so that I can easily reuse it for plotting graphs

Example:

```
# Pivoted data
country  	2005	2006	2007	
Afghanistan	46.6	46.9	47.2	
Albania	        76.1	76.3	76.5	
...

# Unpivoted data
country      year  average_age
Afghanistan  2005  46.6
Afghanistan  2006  46.9
Afghanistan  2007  47.2
Albania      2005  76.1
Albania      2006  76.3
Albania      2007  76.5
...
```

## Acceptance criteria

* [ ] Able to normalize my pivoted data
* [ ] `dpp run un-pivot-processor` provides valid datapackage 
* [ ] Able to pre define the name of columns to insert values

## Tasks

* [x] Do analysis
* [x] Prepare Spec
* [x] Failing Tests for processor
* [x] Create unpivot processor
* [ ] Passing test
* [ ] update REAMDE

## Analysis 

Example Spec for given data:
```
# data
country,2015 income,2015 expenses,2016 income,2016 expenses
xxx,20000,15000,30000,30000
...

# spec
- run: unpivot
  parameters:
    resources: (same as always)
    extraKeyFields:
      -
        name: year
        type: integer
      -
        name: direction
        name: string
    extraValueField:
        name: amount
        type: number
    unpivot:
      - 
        name: 2015 incomes
        keys:
          year: 2015
          direction: In
      - 
        name: 2015 expenses
        keys:
          year: 2015
          direction: Out
      - 
        name: 2016 incomes
        keys:
          year: 2016
          direction: In
      - 
        name: 2016 expenses
        keys:
          year: 2015
          direction: Out
      # or regex 
      -
        name: ([0-9]{4}) (\w+)
        keys:
          year: \1
          direction: \2
```

### PR DESCRIPTION
**Pull request by zelima**
This pull request fixes #108 .

* [x] I've added tests to cover the proposed changes

Changes proposed in this pull request:

* update resource descriptor according to new fields
* yield pivoted row per field name from original row
* support regex to grab column names from original row
* support regex for grabbing value for unpivoted columns
* tests
* instructions in README

For some reasons not able to add @OriHoch as a reviewer

### CODE REVIEW

File: datapackage_pipelines/lib/unpivot.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+        return field_val
+
+
+def unpivot(spec, rows):
+    schema = spec['schema']
+    field_names = list(map(lambda f: f['name'], schema['fields']))
+    for row in rows:
+        for unpivot_field in unpivot_fields_without_regex:
+            new_row = copy.deepcopy(unpivot_field['keys'])
+            for field in field_names:
```
Comment (akariv) [inline datapackage_pipelines/lib/unpivot.py]: why do we need to iterate on all fields here?

File: datapackage_pipelines/lib/unpivot.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+                if field in new_row:
+                    continue
+                from_row = row.get(field)
+                val = row.get(unpivot_field['name'])
+                new_row[field] = from_row or val
```
Comment (akariv) [inline datapackage_pipelines/lib/unpivot.py]: and what if the value in row is `None`? will it always be replaced by `val`?

File: datapackage_pipelines/lib/unpivot.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+            fields_to_pivot = (list(
+                filter(match_fields(field_name_re, True), fields)
+            ))
+            fields = list(
+                filter(match_fields(field_name_re, False), fields)
+            )
+
+            # handle with regex
+            for field_to_pivot in fields_to_pivot:
+                oiginal_key_values = u_field['keys']  # With regex
```
Comment (akariv) [inline datapackage_pipelines/lib/unpivot.py]: *original

File: datapackage_pipelines/lib/unpivot.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+from datapackage_pipelines.wrapper import ingest, spew
+from datapackage_pipelines.utilities.resource_matcher import ResourceMatcher
+
+parameters, datapackage, resource_iterator = ingest()
+
+resources = ResourceMatcher(parameters.get('resources'))
+unpivot_fields = parameters.get('unpivot')
+keys_ = parameters.get('extraKeyFields')
+values_ = parameters.get('extraValueField')
+unpivot_fields_without_regex = []
```
Comment (akariv) [inline datapackage_pipelines/lib/unpivot.py]: please don't use globals like this

File: datapackage_pipelines/lib/unpivot.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+                        field_to_pivot['name'])
+                    # parse value to correct type
+                    new_key_values[key] = parse_field(
+                        keys_, key, new_val)
+                    field_to_pivot['keys'] = new_key_values
+                unpivot_fields_without_regex.append(field_to_pivot)
+
+        fields.extend(keys_)
+        fields.append(values_)
+        resource['schema']['fields'] = fields
```
Comment (akariv) [inline datapackage_pipelines/lib/unpivot.py]: if using setdefault (see above) then unneeded

File: datapackage_pipelines/lib/unpivot.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+            fields = list(
+                filter(match_fields(field_name_re, False), fields)
+            )
+
+            # handle with regex
+            for field_to_pivot in fields_to_pivot:
+                oiginal_key_values = u_field['keys']  # With regex
+                new_key_values = {}
+                for key in oiginal_key_values:
+                    new_val = re.sub(
```
Comment (akariv) [inline datapackage_pipelines/lib/unpivot.py]: Why do we need `parse_field` here?
If it is a string, then we use regexp (and get a string in return).
If it's not, we just copy the value (as obviously it's not a regexp).

File: datapackage_pipelines/lib/unpivot.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                        u_field['name'],
+                        str(oiginal_key_values[key]),
+                        field_to_pivot['name'])
+                    # parse value to correct type
+                    new_key_values[key] = parse_field(
+                        keys_, key, new_val)
+                    field_to_pivot['keys'] = new_key_values
+                unpivot_fields_without_regex.append(field_to_pivot)
+
```
Comment (akariv) [inline datapackage_pipelines/lib/unpivot.py]: at this point `fields` contains the names of fields to keep in each row - you should hold that information and use it later.

File: datapackage_pipelines/lib/unpivot.py (hunk 1 above)
Comment (akariv) [inline datapackage_pipelines/lib/unpivot.py]: In short, for each original row and unpivot_field:
- copy the "keep" fields from the original row to the new row
- copy the "keys" from the unpivot_field
- copy the value of the unpivot_field name from the original row to the new row

File: datapackage_pipelines/lib/unpivot.py (hunk 6 above)
Comment (zelima) [reply datapackage_pipelines/lib/unpivot.py]: @akariv this is the case when it is a string, we use regex, but not want a string in return:
* imagine this is the title of the original field: `incomes 2017` 
* and we want to grab 2017 with regex `\2` from `(\w+) ([0-9]{4})`
* result will be "2017" as string
* but user may need it as number or float etc.. depending type defined in `extraKeyFields`

Comment (akariv) [reply datapackage_pipelines/lib/unpivot.py]: you can let dpp take care of type conversion, and leave it as a string here.

File: datapackage_pipelines/lib/unpivot.py (hunk 5 above)
Comment (zelima) [reply datapackage_pipelines/lib/unpivot.py]: This actually won't work as I'm assigning filtered list of fields on each iteration to [`fields`](https://github.com/frictionlessdata/datapackage-pipelines/pull/110/files#diff-c171b3c088275f83ba688825001b977fR38)
```
fields = [{name: should_remain}, {name: should_go_away}, {name: go_away_as_well}]
for field in fields: 
    fields = fielter(fields, field==should_remain)
```

Meaning `setdefault` loses its sense?

Comment (akariv) [reply datapackage_pipelines/lib/unpivot.py]: you're right

**Comment by zelima** [comment]
@akariv looks like [newest release](https://github.com/wbolster/plyvel/commit/ca3702064ea38d9d7c0be621e12cee581058f586) of `plyvel` (`SPEEDUP_REQUIRES` in setup.py) have broken it's setup. (has not been changed since oct 2016) 
Travis build is failing cause of this. Same with `make install-speedup` locally. Meaning even current build will fail if restarted.

Two ways:
* Fix the version
* use `make install` instead

WDYT?

File: README.md [hunk 8]
```diff
@@ -784,7 +784,10 @@ parameters:
       type: integer
     -
       name: direction
-      name: string
+      type: string
+      enumerate:
```
Comment (akariv) [inline README.md]: See https://frictionlessdata.io/specs/table-schema/#constraints for correct syntax according to tableschema spec

File: README.md [hunk 9]
```diff
@@ -844,7 +848,10 @@ parameters:
...
```
Comment (akariv) [inline README.md]: see above

File: tests/stdlib/fixtures/simple_unpivot [hunk 10]
```diff
@@ -4,7 +4,7 @@ unpivot
   "resources": "balance",
   "extraKeyFields": [
     {"name": "year", "type": "integer"},
-    {"name": "direction", "type": "string"}
+    {"name": "direction", "type": "string", "enumerate": ["In", "Out"]}
```
Comment (akariv) [inline tests/stdlib/fixtures/simple_unpivot]: see above

**Comment by akariv** [comment]
@zelima can you fix the lint error in the tests?

**Comment by zelima** [comment]
@akariv fixed

**Pull request merged**